File: 123. Best Time to Buy and Sell Stock III/solution.py
class Solution:
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        # edge cases
        if len(prices) < 2:
            return 0
        
        # calculate profits from left side
        leftprofits = [0]
        minprice = prices[0]
        for price in prices[1:]:
            minprice = min(price, minprice)
            profit = price - minprice
            maxprofit = max(leftprofits[-1], profit)
            leftprofits.append(profit)

        # calculate profits from right side
        rightprofits = [0]
        maxprice = prices[-1]
        for i in range(len(prices)-2, -1, -1):
            maxprice = max(prices[i], maxprice)
            profit = maxprice - prices[i]
            maxprofit = max(profit, rightprofits[-1])
            rightprofits.append(maxprofit)
        rightprofits.reverse()

        # calculate the maxmimum profits from both sides
        maxprofit = 0
        for i, j in zip(leftprofits, rightprofits):
            maxprofit = max(maxprofit, i+j)
            
        return maxprofit
            
                

File: 123. Best Time to Buy and Sell Stock III/solution.java
class Solution {
    public int maxProfit(int[] prices) {
        if (prices.length < 2) 
            return 0;
        
        int minPrice;
        int maxPrice;
        int profit;
        int[] lp = new int[prices.length];
        int[] rp = new int[prices.length];
        minPrice = prices[0];
        profit = 0;
        
        lp[0] = 0;
        minPrice = prices[0];
        for (int i = 1; i < prices.length; i++) {
            minPrice = Math.min(minPrice, prices[i]);
            profit = prices[i] - minPrice;
            lp[i] = Math.max(profit, lp[i-1]);
        }
        
        rp[prices.length-1] = 0;
        profit = 0;
        maxPrice = prices[prices.length-1];
        for (int i = prices.length - 2; i >= 0; i--) {
            maxPrice = Math.max(maxPrice, prices[i]);
            profit = maxPrice - prices[i];
            rp[i] = Math.max(profit, rp[i+1]);
        }
        
        profit = 0;
        for (int i = 0; i < prices.length; i++) {
            profit = Math.max(profit, lp[i] + rp[i]);
        }
        
        return profit;
    }
}
